runtime/src/bank.rs

```diff
@@ -897,77 +897,116 @@ impl Bank {
-                            self.store_account(&delegation.voter_pubkey, &vote_account);
-
-                            if voters_reward > 0 {
-                                *validator_rewards
-                                    .entry(delegation.voter_pubkey)
-                                    .or_insert(0i64) += voters_reward as i64;
-                            }
+                    (Some(stake_account), Some(vote_account)) => Some((
+                        (*stake_pubkey, stake_account),
+                        (delegation.voter_pubkey, vote_account),
+                    )),
+                    (_, _) => None,
+                }
+            })
+            .collect()
+    }
 
-                            if stakers_reward > 0 {
-                                *validator_rewards.entry(*stake_pubkey).or_insert(0i64) +=
-                                    stakers_reward as i64;
-                            }
+    /// iterate over all stakes, redeem vote credits for each stake we can
+    ///   successfully load and parse, return total payout
+    fn pay_validator_rewards(&mut self, rewards: u64) -> f64 {
+        let stake_history = self.stakes.read().unwrap().history().clone();
 
-                            stakers_reward + voters_reward
-                        } else {
-                            debug!(
-                                "stake_state::redeem_rewards() failed for {}: {:?}",
-                                stake_pubkey, rewards
-                            );
-                            0
-                        }
+        let mut stake_delegation_accounts = self.stake_delegation_accounts();
+
+        let points: u128 = stake_delegation_accounts
+            .iter()
+            .map(
+                |((_stake_pubkey, stake_account), (_vote_pubkey, vote_account))| {
+                    stake_state::calculate_points(
+                        &stake_account,
+                        &vote_account,
+                        Some(&stake_history),
+                    )
+                    .unwrap_or(0)
+                },
+            )
+            .sum();
+
+        if points == 0 {
+            return 0.0;
+        }
+
+        let point_value = PointValue { rewards, points };
+
+        let mut rewards = HashMap::new();
+
+        // pay according to point value
+        stake_delegation_accounts.iter_mut().for_each(
+            |((stake_pubkey, stake_account), (vote_pubkey, vote_account))| {
+                let redeemed = stake_state::redeem_rewards(
+                    stake_account,
+                    vote_account,
+                    &point_value,
+                    Some(&stake_history),
+                );
+                if let Ok((stakers_reward, voters_reward)) = redeemed {
+                    self.store_account(&stake_pubkey, &stake_account);
+                    self.store_account(&vote_pubkey, &vote_account);
```

Comment:
still not sure: when storing vote_pubkey happens multiple times, this overwrites any immediately written account state because we merely get_account at the start of this fn (with `stake_delegation_accounts`) and only once, reusing them. What's the correct behavior?

Reply:
Well, this was another bug... fixed https://github.com/solana-labs/solana/pull/10914/commits/8472fc3f0825c1c2d4a68050bd667095d517081e